guide/src/dev/performance.md

```diff
@@ -0,0 +1,206 @@
+- References to input text avoid copying when possible
+- Heap allocation is only used when necessary for large strings
+
+The `InlineStr` type can store strings up to 22 bytes (on 64-bit systems) directly in the enum variant, avoiding heap allocation for common short string cases. Refer to the [string handling](./string-handling.md) documentation for more details.
+
+## Tree Structure Optimization
+
+The AST (Abstract Syntax Tree) is stored in a vec-based tree structure that provides:
+
+1. Fast node creation during parsing
+2. Efficient tree traversal
+3. Memory locality from vector storage
+
+```rust
+pub(crate) struct Tree<T> {
+    nodes: Vec<Node<T>>,
+    spine: Vec<TreeIndex>,
+    cur: Option<TreeIndex>, 
+}
+
+pub(crate) struct Node<T> {
+    pub child: Option<TreeIndex>,
+    pub next: Option<TreeIndex>,
+    pub item: T,
+}
+```
+
+Key optimizations in the tree structure include:
+
+- Using indices instead of pointers for node references
+- Maintaining a "spine" for fast access to ancestor nodes
+- Storing nodes contiguously in a vector for better cache usage
+- Using non-zero indices to save space in option types
+
+## Single-Pass Parsing
+
+The parser uses a single pass over the input text for both block and inline parsing, avoiding the need to rescan text multiple times. This is achieved through:
+
+1. Maintaining parser state to track context
+2. Handling both block and inline elements incrementally
+3. Using lookahead only when necessary
+
+For example, here's how link reference definitions are handled in a single pass:
+
+```rust
+fn parse_refdef_total(&mut self, start: usize) -> Option<(usize, LinkLabel<'a>, LinkDef<'a>)> {
+    let bytes = &self.text.as_bytes()[start..];
+    if scan_ch(bytes, b'[') == 0 {
+        return None;
+    }
+    let (mut i, label) = self.parse_refdef_label(start + 1)?;
+    i += 1;
+    if scan_ch(&bytes[i..], b':') == 0 {
+        return None;
+    }
+    i += 1;
+    let (bytecount, link_def) = self.scan_refdef(start, start + i)?;
+    Some((bytecount + i, UniCase::new(label), link_def))
+}
+```
+
+## Lazy Text Merging
+
+The library includes a `TextMergeStream` utility that lazily merges consecutive text events:
+
+```rust
+impl<'a, I> Iterator for TextMergeStream<'a, I>
+where
+    I: Iterator<Item = Event<'a>>,
+{
+    type Item = Event<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match (self.last_event.take(), self.iter.next()) {
+            (Some(Event::Text(last)), Some(Event::Text(next))) => {
+                let mut merged = String::from(last);
+                merged.push_str(&next);
+                // Continue merging consecutive text events...
+            }
+            // Handle other cases...
+        }
+    }
+}
+```
+
+This avoids unnecessary string allocations and concatenations when processing text nodes.
+
+## Protection Against Pathological Input
+
+Several protections are in place to prevent exponential time or memory usage on malicious input:
```

Comment:
```suggestion
Several protections are in place to prevent quadratic time or memory usage on malicious input:
```

Reply:
Rewrote to integrate your suggestion WRT emphasizing linear performance:

``` ## Protection Against Pathological Input

It is important that the parser performance remain linear with respect to the input length, otherwise the parser would find itself vulnerable to potential DOS attacks. This may not be important for all consumers, but for anyone depending on the library to handle user generated content this is critical.

Several protections are in place to prevent quadratic time or memory usage on malicious input:```